cmd/cmrel/cmd/validate_gomod.go

```diff
@@ -0,0 +1,353 @@
+		if err != nil {
+			return err
+		}
+
+		base := filepath.Base(path)
+
+		if d.IsDir() {
+			if base == "bin" || base == "_bin" || strings.HasPrefix(base, ".") {
+				return fs.SkipDir
+			}
+
+			return nil
+		}
+
+		if base != "go.mod" {
+			return nil
+		}
+
+		contents, err := os.ReadFile(path)
+		if err != nil {
+			return fmt.Errorf("failed to read module file %q: %s", path, err.Error())
+		}
+
+		f, err := modfile.Parse(path, contents, nil)
+		if err != nil {
+			return fmt.Errorf("failed to parse module file %q: %s", path, err.Error())
+		}
+
+		newMod := internalModule{
+			Name:          f.Module.Mod.Path,
+			LocalRepoPath: strings.TrimPrefix(path, root+"/"),
+			FullGoModPath: path,
+			Module:        f,
+		}
+
+		iml.modules = append(iml.modules, &newMod)
+		iml.internalModuleNames[newMod.Name] = true
+
+		if path == coreModulePath {
+			iml.coreModule = &newMod
+		} else {
+			iml.submodules = append(iml.submodules, &newMod)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if iml.coreModule == nil {
+		return nil, fmt.Errorf("couldn't find and load core module from expected path %q", coreModulePath)
+	}
+
+	iml.replaceMap = make(map[string]module.Version)
+
+	// all core module replacements of third party depedencies should be duplicated for any
+	// submodule, so that the core module is the single source of truth
+	for _, replaceStmt := range iml.coreModule.Module.Replace {
+		iml.replaceMap[replaceStmt.Old.Path] = replaceStmt.New
+	}
+
+	// all submodules should be replaced with local filesystem replacements, so that everything
+	// only builds against what's in the repo
+	for _, submodule := range iml.submodules {
+		iml.replaceMap[submodule.Module.Module.Mod.Path] = module.Version{
+			Path:    filepath.Join(hardLocalReplace, strings.TrimSuffix(submodule.LocalRepoPath, "go.mod")) + "/",
+			Version: "",
+		}
+	}
+
+	// the core module should always be replaced with a filesystem replacement
+	iml.replaceMap[iml.coreModule.Module.Module.Mod.Path] = module.Version{
+		Path:    hardLocalReplace,
+		Version: "",
+	}
+
+	return &iml, nil
+}
+
+// checkCoreModuleReplacements ensures that the core module doesn't have any local replacements which
+// would break third party importers of that module
+func (iml *internalModuleList) checkCoreModuleReplacements() error {
+	var localReplaces []string
+
+	for _, replaceStmt := range iml.coreModule.Module.Replace {
+		if replaceStmt.New.Version == "" {
+			localReplaces = append(localReplaces, replaceStmt.Old.Path)
+		}
+	}
+
+	if len(localReplaces) > 0 {
+		return fmt.Errorf("core module should have no local (filesystem) replaces, but has: %q", strings.Join(localReplaces, ", "))
+	}
+
+	return nil
+}
+
+// checkReplaces verifies that all internal modules have valid replace statements, meaning that:
+// - if a replace statement is for an internal module, it's using a filesystem replacement
+// - internal modules require a hardcoded dummy version
```

Comment:
this is done in another function I think

Reply:
Good spot, thanks!